### DISCUSSION TITLE
CDR Idea: explicit jump table

### DISCUSSION THREAD

**Discussion by Gleefre** [fstamour/cl-car#22] [Ideas]
"Special form for optimized `(case X (0 ...) (1 ...) (2 ...) ...)`"

There is still no discussion format; which I think can be fixed by starting doing something (and then it will hopefully become a necessity at some point, which will hopefully "make it happen")

An idea from a hacker news thread here: https://news.ycombinator.com/item?id=20789547

It is not possible (or, at least, it is very tricky -- I don't know any possible solution) to write a portable macro that would allow to execute one of N branches based on its index.

Some implementations optimize `case` with clauses 0, 1, ..., N-1, for example SBCL and CCL seems to be doing this (but I didn't test it thoroughly).

I don't have any good names for such a special form yet x/x

**Comment by fstamour** [comment]
I think it could be called "jump table".

Since it's something that an implementation can already optimize for, I'm not sure what a CDR for this could provide?

Perhaps it might be nice to document to possible optimizations an implementation could do...

**Comment by Gleefre** [comment]
A few possibilities:

1) Given the CDR 14 (about detecting implementation support for CDRs), it could become possible to reliably check if `case` is using the jump-table optimization.

2) Even though `case` is a macro, it is not possible to implement it "efficiently" (that is, with the jump table optimization) in CL right now using already existing special forms.

3) Using `case` for the jump table optimization is not ideal, as it would need to re-check if the jump-table optimization is applicable; and depending on the optimization settings / execution mode the optimization might not be used at all. So it would become possible to use the jump-table optimization when writing other constructs (like switch) somewhat more reliably and without needing to use `case`.
